### ISSUE TITLE
Firing alerts go back to pending on owner pod recreation

### ISSUE THREAD

**Issue by oneacl** [grafana/mimir#3814]
#### Describe the bug

I run mimir 2.5.0 installed via helm (the cortex reference in my code below is due to the name override as it was a cortex cluster migrated to mimir a while back).

I have 3 ruler replicas and if the pod owning the rule gets recreated, the rule goes to pending mode across all replica pods.

#### To Reproduce

```
$ date;p|grep ruler
Fri Dec 23 10:38:48 AM UTC 2022
cortex-ruler-66c6bc4b4b-b74lt                    1/1     Running     0             9m20s   10.51.1.117    l1      <none>           <none>
cortex-ruler-66c6bc4b4b-fgvlw                    1/1     Running     0             8m26s   10.51.9.126    l3      <none>           <none>
cortex-ruler-66c6bc4b4b-nvdsk                    1/1     Running     0             4m8s    10.51.7.105    l4      <none>           <none>
$ 

$ cat ruler-alerts
curl -s "$1":8080/prometheus/api/v1/alerts -H 'X-Scope-OrgId: prom-helm'|jq .data.alerts[].state -r
$ ./ruler-alerts 10.51.1.117
firing
$ ./ruler-alerts 10.51.9.126
firing
$ ./ruler-alerts 10.51.7.105
firing
$
```

```
$ k logs -f cortex-ruler-66c6bc4b4b-fgvlw
level=debug ts=2022-12-23T10:40:27.586310136Z caller=ruler.go:429 msg="syncing rules" reason=periodic
level=debug ts=2022-12-23T10:40:27.714630101Z caller=ruler.go:566 msg="rule group not owned, ignoring" user=prom-helm namespace=test1ns name=test1group


$ k logs -f cortex-ruler-66c6bc4b4b-b74lt
level=debug ts=2022-12-23T10:41:34.48743144Z caller=ruler.go:429 msg="syncing rules" reason=periodic
level=debug ts=2022-12-23T10:41:34.6179574Z caller=ruler.go:566 msg="rule group not owned, ignoring" user=prom-helm namespace=test1ns name=test1group

$ k logs -f cortex-ruler-66c6bc4b4b-nvdsk
level=debug ts=2022-12-23T10:40:48.886941147Z caller=ruler.go:429 msg="syncing rules" reason=periodic
level=debug ts=2022-12-23T10:40:49.195480065Z caller=ruler.go:563 msg="rule group owned" user=prom-helm namespace=test1ns name=test1group
level=debug ts=2022-12-23T10:40:49.259731668Z caller=manager.go:155 msg="rules have not changed, skipping rule manager update" user=prom-helm
```

Delete the pod owning the rules:

```
$ k delete po cortex-ruler-66c6bc4b4b-nvdsk
$ date;p|grep ruler
Fri Dec 23 10:42:34 AM UTC 2022
cortex-ruler-66c6bc4b4b-b74lt                    1/1     Running     0             13m     10.51.1.117    l1      <none>           <none>
cortex-ruler-66c6bc4b4b-fgvlw                    1/1     Running     0             12m     10.51.9.126    l3      <none>           <none>
cortex-ruler-66c6bc4b4b-wxmmf                    0/1     Running     0             26s     10.51.7.106    l4      <none>           <none>
$
```

An instance picks up the rules due to the ring change:

```
$ k logs -f cortex-ruler-66c6bc4b4b-fgvlw
level=debug ts=2022-12-23T10:42:14.044946167Z caller=ruler.go:429 msg="syncing rules" reason=ring-change
level=debug ts=2022-12-23T10:42:14.175042671Z caller=ruler.go:563 msg="rule group owned" user=prom-helm namespace=test1ns name=test1group
level=info ts=2022-12-23T10:42:14.240412926Z caller=mapper.go:166 msg="updating rule file" file=/data/prom-helm/test1ns
level=debug ts=2022-12-23T10:42:14.242123745Z caller=manager.go:195 msg="creating rule manager for user" user=prom-helm
level=debug ts=2022-12-23T10:42:14.244221447Z caller=manager.go:159 msg="updating rules" user=prom-helm
level=info ts=2022-12-23T10:42:14.249007578Z caller=manager.go:998 user=prom-helm msg="Starting rule manager..."


$ k logs -f cortex-ruler-66c6bc4b4b-b74lt
level=warn ts=2022-12-23T10:42:24.488181903Z caller=pool.go:184 msg="removing ruler failing healthcheck" addr=10.51.7.105:9095 reason="rpc error: code = DeadlineExceeded desc = context deadline exceeded"
level=debug ts=2022-12-23T10:42:34.487387435Z caller=ruler.go:429 msg="syncing rules" reason=periodic
level=debug ts=2022-12-23T10:42:34.697138281Z caller=ruler.go:566 msg="rule group not owned, ignoring" user=prom-helm namespace=test1ns name=test1group
```

All instances now go to pending mode:

```
$ ./ruler-alerts 10.51.1.117
pending
$ ./ruler-alerts 10.51.9.126
pending
$ ./ruler-alerts 10.51.7.106
pending
$
```

This specific rule has a 5m 'for' period - and while the rule is in pending mode - alertmanager reaches the "endsAt" time causing resolutions to be sent - and a couple of minutes later the 'firing' alert goes through again.

In my case the 'resolution' is causing issues downstream (as it calls a workflow).

My diff-config is fairly simple:
```
ruler:
    alertmanager_url: dnssrvnoa+http://_http-metrics._tcp.cortex-alertmanager-headless.default.svc.cluster.local/alertmanager
    rule_path: /data
ruler_storage:
    backend: s3
    s3:
        access_key_id: ***
        bucket_name: alertmanager
        secret_access_key: '********'
    storage_prefix: ruler
```

Is it possible to have the state stored in S3? (I see only the rule definition is currently kept there) - or some other way to keep the alert state replicated across the ring when the 'owner' pod goes away and another one picks it up?

I can't seem to find anything in the documentation, the only relevant parts would be for_outage_tolerance (set at default 1h) and for_grace_period (also at default 10m).

#### Environment

- Infrastructure: Kubernetes
- Deployment tool: helm

Thank you.

**Comment by oneacl** [comment]
As a follow-up (related) question - would it be possible for the ruler to send a custom 'endsAt' time to be at least a few days in the future? this would (in theory) mean that during a ruler outage alertmanager would not send a resolution and when a new ruler comes online it would re-evaluate and after the 'for' period it would change to a 'firing' state again.

Looking at the source code most 'endsAt' come from the prometheus/alertmanager vendor.

Thanks.

**Comment by dimitarvdimitrov** [comment]
nice investigation! thank you for the great details. This looks very similar to https://github.com/grafana/mimir/issues/2888. What we did to solve this internally was to set `for_grace_period: 2m`. Can you try setting that to a duration lower than the eval period of your rule group?

In the meantime I will open a PR to change the default to 2m

**Comment by oneacl** [comment]
Thank you very much for the quick reply and the detailed update of the docs in your PR - really appreciate it.
The updated docs definitely make more sense to me now.

I updated the for_grace_period to both 2m and 0s and it appears to work as expected. The reason for setting it to 0s is that I want all alerts to have their state restored - I can't think of a reason where I wouldn't want them restored.

Based on a few tests, the timeline appears to be:
- delete the ruler pod that owns the rules
- another pod picks up the rules
- for about 1 minute the API returns no alerts (and Grafana shows the rule as 'Normal')
- after said minute the rules go into Pending mode and the debug message about the rules restored shows:
- `msg="'for' state restored" alertname=test1`
- the rules will stay in Pending mode for 2 more minutes (approximately) then go into firing mode

No resolution is being sent but I'm worried that the 3 minutes observed here are very close to the endsAt value that prometheus sets for the alert and since nothing is being sent there's a likelihood that alertmanager will resolve the alert (and ruler will fire it shortly after).

Another issue would be a ruler outage for > 3 minutes which will then cause all alertmanager alerts to resolve (reaching their endsAt value).

What I think would complement this setting is the ability to send a very large value for endsAt (maybe upwards of 24h or at least a value that one can comfortably do ruler "maintenance" without sending out resolutions).

Would this be something that would be possible to do?

Thanks again - really appreciate your help.

**Comment by dimitarvdimitrov** [comment]
> No resolution is being sent but I'm worried that the 3 minutes observed here are very close to the endsAt value that prometheus sets for the alert and since nothing is being sent there's a likelihood that alertmanager will resolve the alert (and ruler will fire it shortly after).
> 
> Another issue would be a ruler outage for > 3 minutes which will then cause all alertmanager alerts to resolve (reaching their endsAt value).

The `endsAt` parameters that's being sent to the alertmanager is 4 times the alert "for" duration or 4 times the resend delay. If you increase the `-ruler.resend-delay` to `6h`, then the alertmanager will not autoresolve the alerts for 24h. This is the code 

https://github.com/grafana/mimir/blob/31d9b85cd642ebb3edc3bd6ae3edc59e336d0d42/vendor/github.com/prometheus/prometheus/rules/alerting.go#L504-L508

We should probably document this behaviour.

**Comment by changliu-wk** [comment]
@dimitarvdimitrov 
I think this is related, I have an alert is firing, and its metrics like this: 

![Image](https://github.com/user-attachments/assets/49ed3915-3392-4c2c-ba0d-b453a5266ad3)

I am confused why the `ALERTS` metric sometimes has a gap, sometimes not, during pod switching due to node eviction. I dug around the code, found it won't store the `ALERTS` metric until [restore](https://github.com/grafana/mimir/blob/0c1e06f13b90434e0e4da6118dcc2191ea23e8d7/vendor/github.com/prometheus/prometheus/rules/alerting.go#L485-L486) is true. It is supposed to be broken, but why it could be continuous sometime? Is it because this part of [code](https://github.com/grafana/mimir/blob/0c1e06f13b90434e0e4da6118dcc2191ea23e8d7/vendor/github.com/prometheus/prometheus/rules/group.go#L267)? The old pod send out the ALERTS metric to make it not broken. Any help is appreciated, thanks!

**Comment by dimitarvdimitrov** [comment]
`ALERTS` contains one sample every time an alert evaluation is considered firing. If the ruler restarts are skips an evaluation, then there is no sample. For example, if the ruler restarts 1s before the next evaluation and the new pod takes longer than 1s to start, then one evaluation will be skipped. Depending on how long shutting down and starting up take, there may be more evaluations skipped. 

We did a change (https://github.com/prometheus/prometheus/pull/15804) which should significantly reduce shutdown times for rulers with many rule groups. This change is included in 2.16.0
